typing/typecore.ml

```diff
@@ -4999,10 +5004,16 @@ and type_construct env loc lid sarg ty_expected_explained attrs =
 (* Typing of statements (expressions whose values are discarded) *)
 
 and type_statement ?explanation env sexp =
+  let allow_polymorphic e = match e.pexp_desc with
+    | Pexp_while _ -> true
```

Comment:
I would prefer if we had a code comment that explains why we are special-casing this construct. In particular, the comment should explicitly mention the version where it was added, so that we can decide (when we meet this part of the code again in N years) whether it is time to remove the special case.